src/Spectre.libPeakFinder/FWHHCalculator.cpp

```diff
@@ -0,0 +1,92 @@
+/*
+* FWHHCalculator.cpp
+* Class for calculating locations of full width at half height markers for
+* input peak data.
+*
+Copyright 2018 Dariusz Kuchta
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#include "FWHHCalculator.h"
+#include "Spectre.libFunctional/Transform.h"
+
+namespace spectre::algorithm::peakfinder
+{
+    Signal FWHHCalculator::GetLeftFWHH(const SignalView x, const SignalView y, const IndicesView valleys, const IndicesView peaks)
+    {
+        Signal result;
+        result.resize(peaks.size());
+
+        for (int i = 0; i < peaks.size(); ++i)
+        {
+            Index peakIdx = peaks[i];
+            Index valleyIdx = valleys[i];
+            DataType peakVal = y[peakIdx];
+
+            SignalView segment = y.subspan(valleyIdx, peakIdx - valleyIdx);
+            DataType lfwhhPeakVal = peakVal * 0.5;
+            Signal segmentNegated = spectre::core::functional::transform<DataType>(segment, [](DataType val) { return -val; });
+            Index lfwhhNearestIdx = GetClosestNeighbourIndex(segmentNegated, -lfwhhPeakVal) + valleyIdx;
+            DataType lfwhhVal = Lerp(y[lfwhhNearestIdx], y[lfwhhNearestIdx + 1], x[lfwhhNearestIdx], x[lfwhhNearestIdx + 1], lfwhhPeakVal);
+
+            result[i] = lfwhhVal;
```

Comment:
What about extracting calculation of single `result[i]` to an `inline` function of something and just running `transform`?

Reply:
Totally agree with extracting.

Reply:
This is very hard to do without doing lots of branching.

Reply:
Well, no. You are checking only against the end of the collection. The rest of the code is linear. There is no branching.